### ISSUE TITLE
Configurable key bindings - Requests and Discussion

### ISSUE THREAD

**Issue by SomeTroglodyte** [yairm210/Unciv#8862]
## Context

Unciv may get configurable keyboard key bindings in the near future (PR #8802). Since there are **many** functions that may be candidates for such a binding, implementing them all is a work in progress and might take a long time. Please bundle any and all requests / comments / criticism on that feature here in this issue.

## Goal

Maintain and prioritize a list of desired functions to get a configurable key binding.
Possibly discover usages we might otherwise miss.

## Status

* [x] Access to configuration gated by switch on "Debug" page
* [ ] Has open PRs

## Framework stuff
* Interoperability with reusable components (API level)
    * [x] Popup
    * [x] WASD Listener
    * [x] onActivation
    * [ ] TabbedPager
    * [x] UncivTooltip
    * [x] ExpanderTab?
    * [ ] UncivSlider?
    * [ ] Civilopedia migration to TabbedPager?
    * [x] UnitActions are dependent on the "Show more" state, but should not be

## List of Contexts / Bindings

*   [x] Main Menu
    *   [x] Resume, Quickstart, New, Load
    *   [x] Editor, Multiplayer, Mods, Options
    *   [x] Civilopedia - same binding as WorldScreen
*   [x] World Screen
    *   [x] Next turn with alternate
    *   [x] Civilopedia
    *   [x] Empire Overview
    *   [x] F2-F10 Various info screens
    *   [x] F11 F12 Quicksave/Load
    *   [x] Options, Save, Load, Toggle UI
    *   [x] Toogle display (Resources, Yields, Worked, Movements)
    *   [x] Zoom In/Out
    *   [x] Panning with WASD and Arrow keys
    *   [x] Show Capital (Home)
    *   [ ] ~Multiselect (Shift) - Not straightforward~ Nope won't do
*   [x] Unit Actions
    *   [x] All of them as far as I know
*   [x] City Screen
    *   [x] Left/Right arrows
    *   [x] Buy construction, Buy tile
    *   [x] Stats breakdown (new)
    *   [x] .. new ones for the top right expanders open/close??
    *   [x] .. new ones for switching production focus??
    *   [x] .. new ones for opening+jumping-to the construction categories (left-side expanders)??
*   [ ] Popups
    *   [x] Confirm / Reject types of Dialog
        *   [x] Confirm / Yes
        *   [x] Reject / No
    *   [ ] Options Popup
        *   [ ] F12 - dev option - likely won't do
        *   [ ] New bindings to navigate tabs? - see TabbedPager mention
    *   [ ] Conquered city
        *   [ ] Annex
        *   [ ] Destroy
        *   [ ] Liberate
        *   [ ] Puppet
        *   [ ] Raze
*   [ ] Picker Screens
    *   [ ] Load / Save
    *   [ ] Improvement picker - how to deal with improvement json `shortcutKey`?
    *   [ ] Change any general Picker Screen Confirm/Cancel stuff?
    *   [ ] ... ?
*   [ ] Civilopedia
    * Arrow Navigation (↔ Categories, ↕ Entries)
    * Category selection by Letter (are the current keys translation-dependent?)

**Comment by Gualdimar** [comment]
I just tried the button assignment tab in the settings. It looks a bit complicated😵 In terms of user-friendly experience (all these drop-downs, along with hovering). The drop-down lists contain characters from gamepads and shift+key characters (although you can't assign them via hover)

Didn't you think of something more common and simple?
For example, just a non-editible text field, which indicates the current binding. When a user clicks on the field, it is highlighted in a different color, and then:
either the user clicks ESC and the binding does not change
or the user presses some button and it is binded to the corresponding action.

**Comment by Gualdimar** [comment]
And here is a list of standard assignments(from civilopedia):
Space or N - Next unit or turn
E - Empire overview (last viewed page)
\+ - Zoom in
\- - Zoom out
Home - center on capital or open its city screen if already centered
F1 - Open Civilopedia
F2 - Empire overview Trades
F3 - Empire overview Units
F4 - Empire overview Diplomacy
F5 - Social policies
F6 - Technologies
F7 - Empire overview Cities
F8 - Victory Progress
F9 - Empire overview Stats
F10 - Empire overview Resources
F11 - Quicksave
F12 - Quickload
Ctrl-R - Toggle tile resource display
Ctrl-Y - Toggle tile yield display
Ctrl-O - Game options
Ctrl-S - Save game
Ctrl-L - Load game
Ctrl-U - Toggle UI (World Screen only)

**Comment by Gualdimar** [comment]
> Another json listing allowed codes? Nah. waiting for better idea.

That's why I suggest to abandon the drop-down lists in favor of a text field (which will display only keys that the user can actually press)

> Since all that is relatively encapsulated, no biggie to change.

I'm not speaking from the point of view of a coder or the complexity of the code. I'm sharing my impressions of using the interface, based on my experience and habits from other games.



> Physical keys as they would look on an Imperial (sorry, en-US) layout. Virtual keys.

There are keyboards with separate buttons for @,# and : ? Hm...🤔

> one or two you missed

My list based on the results of the Input.Keys.* search, I haven't seen the ones you mentioned.

And on the latest version, an exception when opening Options:
<details>

```
java.lang.ExceptionInInitializerError
	at com.unciv.ui.popups.options.KeyBindingsTab$KeyboardBindingWidget.<init>(KeyBindingsTab.kt:74)
	at com.unciv.ui.popups.options.KeyBindingsTab.<init>(KeyBindingsTab.kt:42)
	at com.unciv.ui.popups.options.OptionsPopup.showOrHideKeyBindings$core(OptionsPopup.kt:193)
	at com.unciv.ui.popups.options.OptionsPopup.<init>(OptionsPopup.kt:125)
	at com.unciv.ui.screens.basescreen.BaseScreen.openOptionsPopup(BaseScreen.kt:166)
	at com.unciv.ui.screens.basescreen.BaseScreen.openOptionsPopup$default(BaseScreen.kt:165)
	at com.unciv.ui.screens.mainmenuscreen.MainMenuScreen$optionsTable$1.invoke(MainMenuScreen.kt:187)
	at com.unciv.ui.screens.mainmenuscreen.MainMenuScreen$optionsTable$1.invoke(MainMenuScreen.kt:186)
	at com.unciv.ui.components.extensions.Scene2dExtensionsKt$onActivation$1.invoke(Scene2dExtensions.kt:190)
	at com.unciv.ui.components.extensions.Scene2dExtensionsKt$onActivation$1.invoke(Scene2dExtensions.kt:188)
	at com.unciv.ui.components.extensions.ActorAttachments.activate(Scene2dExtensions.kt:136)
	at com.unciv.ui.components.extensions.Scene2dExtensionsKt.activate(Scene2dExtensions.kt:179)
	at com.unciv.ui.components.extensions.ActorAttachments$addActivationAction$2.clicked(Scene2dExtensions.kt:147)
	at com.badlogic.gdx.scenes.scene2d.utils.ClickListener.touchUp(ClickListener.java:88)
	at com.badlogic.gdx.scenes.scene2d.InputListener.handle(InputListener.java:71)
	at com.badlogic.gdx.scenes.scene2d.Stage.touchUp(Stage.java:355)
	at com.unciv.ui.screens.basescreen.UncivStage.access$touchUp$s80204510(UncivStage.kt:18)
	at com.unciv.ui.screens.basescreen.UncivStage$touchUp$1.invoke(UncivStage.kt:90)
	at com.unciv.ui.screens.basescreen.UncivStage$touchUp$1.invoke(UncivStage.kt:90)
	at com.unciv.ui.crashhandling.CrashHandlingExtensionsKt$wrapCrashHandling$1.invoke(CrashHandlingExtensions.kt:17)
	at com.unciv.ui.screens.basescreen.UncivStage.touchUp(UncivStage.kt:90)
	at com.badlogic.gdx.InputEventQueue.drain(InputEventQueue.java:70)
	at com.badlogic.gdx.backends.lwjgl3.DefaultLwjgl3Input.update(DefaultLwjgl3Input.java:189)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update(Lwjgl3Window.java:378)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:192)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:166)
	at com.unciv.app.desktop.DesktopLauncher.main(DesktopLauncher.kt:76)
Caused by: java.lang.NullPointerException: toString(keyCode) must not be null
	at com.unciv.ui.components.extensions.GdxKeyCodeFixes.toString(Scene2dExtensions.kt:578)
	at com.unciv.ui.components.KeysSelectBox.<clinit>(KeysSelectBox.kt:49)
	... 27 more

```
</details>

**Comment by Gualdimar** [comment]
> Did you find them in code? I missed them and searched in vain - if they're on the actor itself, then the current next PR is in error.

Nah, this was taken from civiloledia.
I checked in version 4.4.9 these combinations do not work, so I was tricked...

**Comment by Gualdimar** [comment]
> Yair removes them

he deleted everything related to the buttons on the minimap in dbc3bcb5a1c956ce0e5dccc2c255562dad411d35, moreover, you thanked him😀
![image](https://user-images.githubusercontent.com/1225948/227002490-24201ede-e10d-411d-9e90-96cfcb58053f.png)

**Comment by saud2410** [comment]
Can you include `Esc` key and `game shutdown popup` keybind?
On 4.10.11 win64, Esc cannot be assigned for other keybind.
And there is no "Game shutdown" entry on settings - key.
I want to use Esc for main menu key on world screen.

**Comment by SeventhM** [comment]
Since you're probably getting a notification anyways, might as well ping. @SomeTroglodyte if this is done or something you can close this. Give the more generic nature of this as key bindings discussion, I'll give it a label

**Comment by saud2410** [comment]
Can we have `disbinding keys`? Functions i want are not much, and good key positions are even more limited. I don't wanna assign keybinds on any other unnecessary functions.

+ And it seems we need `Reset all at once to default` button or something. When i wanna reset keybinds, i have to find and double-click everything.

**Comment by SomeTroglodyte** [comment]
> Can we have

I vaguely remember had both implemented...

Edit: #11476

**Comment by wonderinglostsoul44** [comment]
<html>
<body>
<p data-start="0" data-end="381">Below is a long‑form comment you can paste into #8862.<br data-start="54" data-end="57">
It (i) reframes the problem as an <strong data-start="91" data-end="119">input‑layer architecture</strong> question, (ii) supplies an incremental migration path that avoids regressions on Android, desktop, and WebGL, and (iii) drops in a ready‑made “wish‑list” of bindings plus some <em data-start="296" data-end="314">power‑user candy</em> that could make Unciv’s controls best‑in‑class for open‑source 4X.</p>
<hr data-start="383" data-end="386">
<h2 data-start="388" data-end="453">1  Zooming‑out – treat key‑binding as an “Input&nbsp;Action” system</h2>
<p data-start="455" data-end="560">Instead of sprinkling <code data-start="477" data-end="490">onKeyDown()</code> checks through UI widgets, wrap everything in a <strong data-start="539" data-end="559">three‑tier model</strong>:</p>
<div class="_tableContainer_16hzy_1"><div tabindex="-1" class="_tableWrapper_16hzy_14 group flex w-fit flex-col-reverse">
Tier | Role | Key classes / objects
-- | -- | --
Action | Game‑semantics (NextTurn, QuickSave, OpenCivilopedia) | enum class Action { … }
Binding | <kbd>KeyCombo</kbd> → Action in a Context | data class Binding(val context: Ctx, val combo: KeyCombo, val action: Action)
Resolver | Picks the first active binding under current context stack; fires event | InputRouter singleton

<div class="sticky end-(--thread-content-margin) h-0 self-end select-none"><div class="absolute end-0 flex items-end"><span class="" data-state="closed"><button class="bg-token-bg-primary hover:bg-token-bg-tertiary text-token-text-secondary my-1 rounded-sm p-1 transition-opacity group-[:not(:hover):not(:focus-within)]:pointer-events-none group-[:not(:hover):not(:focus-within)]:opacity-0"><svg width="20" height="20" viewBox="0 0 20 20" fill="currentColor" xmlns="http://www.w3.org/2000/svg" class="icon"><path d="M12.668 10.667C12.668 9.95614 12.668 9.46258 12.6367 9.0791C12.6137 8.79732 12.5758 8.60761 12.5244 8.46387L12.4688 8.33399C12.3148 8.03193 12.0803 7.77885 11.793 7.60254L11.666 7.53125C11.508 7.45087 11.2963 7.39395 10.9209 7.36328C10.5374 7.33197 10.0439 7.33203 9.33301 7.33203H6.5C5.78896 7.33203 5.29563 7.33195 4.91211 7.36328C4.63016 7.38632 4.44065 7.42413 4.29688 7.47559L4.16699 7.53125C3.86488 7.68518 3.61186 7.9196 3.43555 8.20703L3.36524 8.33399C3.28478 8.49198 3.22795 8.70352 3.19727 9.0791C3.16595 9.46259 3.16504 9.95611 3.16504 10.667V13.5C3.16504 14.211 3.16593 14.7044 3.19727 15.0879C3.22797 15.4636 3.28473 15.675 3.36524 15.833L3.43555 15.959C3.61186 16.2466 3.86474 16.4807 4.16699 16.6348L4.29688 16.6914C4.44063 16.7428 4.63025 16.7797 4.91211 16.8027C5.29563 16.8341 5.78896 16.835 6.5 16.835H9.33301C10.0439 16.835 10.5374 16.8341 10.9209 16.8027C11.2965 16.772 11.508 16.7152 11.666 16.6348L11.793 16.5645C12.0804 16.3881 12.3148 16.1351 12.4688 15.833L12.5244 15.7031C12.5759 15.5594 12.6137 15.3698 12.6367 15.0879C12.6681 14.7044 12.668 14.211 12.668 13.5V10.667ZM13.998 12.665C14.4528 12.6634 14.8011 12.6602 15.0879 12.6367C15.4635 12.606 15.675 12.5492 15.833 12.4688L15.959 12.3975C16.2466 12.2211 16.4808 11.9682 16.6348 11.666L16.6914 11.5361C16.7428 11.3924 16.7797 11.2026 16.8027 10.9209C16.8341 10.5374 16.835 10.0439 16.835 9.33301V6.5C16.835 5.78896 16.8341 5.29563 16.8027 4.91211C16.7797 4.63025 16.7428 4.44063 16.6914 4.29688L16.6348 4.16699C16.4807 3.86474 16.2466 3.61186 15.959 3.43555L15.833 3.36524C15.675 3.28473 15.4636 3.22797 15.0879 3.19727C14.7044 3.16593 14.211 3.16504 13.5 3.16504H10.667C9.9561 3.16504 9.46259 3.16595 9.0791 3.19727C8.79739 3.22028 8.6076 3.2572 8.46387 3.30859L8.33399 3.36524C8.03176 3.51923 7.77886 3.75343 7.60254 4.04102L7.53125 4.16699C7.4508 4.32498 7.39397 4.53655 7.36328 4.91211C7.33985 5.19893 7.33562 5.54719 7.33399 6.00195H9.33301C10.022 6.00195 10.5791 6.00131 11.0293 6.03809C11.4873 6.07551 11.8937 6.15471 12.2705 6.34668L12.4883 6.46875C12.984 6.7728 13.3878 7.20854 13.6533 7.72949L13.7197 7.87207C13.8642 8.20859 13.9292 8.56974 13.9619 8.9707C13.9987 9.42092 13.998 9.97799 13.998 10.667V12.665ZM18.165 9.33301C18.165 10.022 18.1657 10.5791 18.1289 11.0293C18.0961 11.4302 18.0311 11.7914 17.8867 12.1279L17.8203 12.2705C17.5549 12.7914 17.1509 13.2272 16.6553 13.5313L16.4365 13.6533C16.0599 13.8452 15.6541 13.9245 15.1963 13.9619C14.8593 13.9895 14.4624 13.9935 13.9951 13.9951C13.9935 14.4624 13.9895 14.8593 13.9619 15.1963C13.9292 15.597 13.864 15.9576 13.7197 16.2939L13.6533 16.4365C13.3878 16.9576 12.9841 17.3941 12.4883 17.6982L12.2705 17.8203C11.8937 18.0123 11.4873 18.0915 11.0293 18.1289C10.5791 18.1657 10.022 18.165 9.33301 18.165H6.5C5.81091 18.165 5.25395 18.1657 4.80371 18.1289C4.40306 18.0962 4.04235 18.031 3.70606 17.8867L3.56348 17.8203C3.04244 17.5548 2.60585 17.151 2.30176 16.6553L2.17969 16.4365C1.98788 16.0599 1.90851 15.6541 1.87109 15.1963C1.83431 14.746 1.83496 14.1891 1.83496 13.5V10.667C1.83496 9.978 1.83432 9.42091 1.87109 8.9707C1.90851 8.5127 1.98772 8.10625 2.17969 7.72949L2.30176 7.51172C2.60586 7.0159 3.04236 6.6122 3.56348 6.34668L3.70606 6.28027C4.04237 6.136 4.40303 6.07083 4.80371 6.03809C5.14051 6.01057 5.53708 6.00551 6.00391 6.00391C6.00551 5.53708 6.01057 5.14051 6.03809 4.80371C6.0755 4.34588 6.15483 3.94012 6.34668 3.56348L6.46875 3.34473C6.77282 2.84912 7.20856 2.44514 7.72949 2.17969L7.87207 2.11328C8.20855 1.96886 8.56979 1.90385 8.9707 1.87109C9.42091 1.83432 9.978 1.83496 10.667 1.83496H13.5C14.1891 1.83496 14.746 1.83431 15.1963 1.87109C15.6541 1.90851 16.0599 1.98788 16.4365 2.17969L16.6553 2.30176C17.151 2.60585 17.5548 3.04244 17.8203 3.56348L17.8867 3.70606C18.031 4.04235 18.0962 4.40306 18.1289 4.80371C18.1657 5.25395 18.165 5.81091 18.165 6.5V9.33301Z"></path></svg></button></span></div></div></div></div>
<p data-start="5744" data-end="5892">Add CI job that loads <strong data-start="5766" data-end="5802">player‑exported binding profiles</strong> found in <code data-start="5812" data-end="5825">./bindings/</code> and validates there are <em data-start="5850" data-end="5871">no duplicate combos</em> within each context.</p>
<hr data-start="5894" data-end="5897">
<h2 data-start="5899" data-end="5924">8&nbsp;&nbsp;Migration &amp; rollout</h2>
<ol data-start="5926" data-end="6372">
<li data-start="5926" data-end="6004">
<p data-start="5929" data-end="6004"><strong data-start="5929" data-end="5946">Phase&nbsp;0 (now)</strong> – skeleton API behind “Debug → Keybinding beta” toggle.</p>
</li>
<li data-start="6005" data-end="6145">
<p data-start="6008" data-end="6145"><strong data-start="6008" data-end="6019">Phase&nbsp;1</strong> – ship World‑screen actions + JSON import/export; ask community to contribute profiles for AZERTY, QWERTZ, Dvorak, Colemak.</p>
</li>
<li data-start="6146" data-end="6249">
<p data-start="6149" data-end="6249"><strong data-start="6149" data-end="6160">Phase&nbsp;2</strong> – port City, Picker, Dialog contexts; deprecate hard‑coded <code data-start="6220" data-end="6240">isKeyJustPressed()</code> calls.</p>
</li>
<li data-start="6250" data-end="6372">
<p data-start="6253" data-end="6372"><strong data-start="6253" data-end="6264">Phase&nbsp;3</strong> – enable on Android with <em data-start="6290" data-end="6302">long‑press</em> and <em data-start="6307" data-end="6319">double‑tap</em> gesture recording; announce in Play&nbsp;Store changelog.</p>
</li>
</ol>
<hr data-start="6374" data-end="6377">
<h3 data-start="6379" data-end="6399">Mic-drop summary</h3>
<blockquote data-start="6401" data-end="6564">
<p data-start="6403" data-end="6564"><strong data-start="6403" data-end="6562">Turn one of Unciv’s roughest UX edges into a moddable, profile‑driven input system that rivals AAA 4X titles—without breaking Android or mod compatibility.</strong></p>
</blockquote>
<p data-start="6566" data-end="6797">Implementing Actions + Bindings + Resolver once lets every screen, unit, and future mod share the same rock‑solid foundation—and makes the community your QA force by letting them export, swap, and perfect their own control schemes.</p>
</body>
</html>Below is a long‑form comment you can paste into #8862.
It (i) reframes the problem as an input‑layer architecture question, (ii) supplies an incremental migration path that avoids regressions on Android, desktop, and WebGL, and (iii) drops in a ready‑made “wish‑list” of bindings plus some power‑user candy that could make Unciv’s controls best‑in‑class for open‑source 4X.

1  Zooming‑out – treat key‑binding as an “Input Action” system
Instead of sprinkling onKeyDown() checks through UI widgets, wrap everything in a three‑tier model:

Tier	Role	Key classes / objects
Action	Game‑semantics (NextTurn, QuickSave, OpenCivilopedia)	enum class Action { … }
Binding	<kbd>KeyCombo</kbd> → Action in a Context	data class Binding(val context: Ctx, val combo: KeyCombo, val action: Action)
Resolver	Picks the first active binding under current context stack; fires event	InputRouter singleton

Benefits:

Conflict‑free – one resolver = one place to handle priority & bubbling.

Re‑usable – gamepad, mouse, touch “gestures” become just more KeyCombo providers.

Mod‑friendly – mods register new Actions and let the user map them through the same UI.

(If you want prior art: Unity’s new Input System or Godot’s InputEventAction layer work exactly like this – proven in cross‑platform titles.)

2  File‑format proposal (keybindings.json)
json
Copy
{
  "version": 2,
  "meta": { "locale": "en-US", "exported": "2025-06-21T18:40Z" },
  "contexts": {
    "Global": [
      { "combo": "Ctrl+S",   "action": "QuickSave"  },
      { "combo": "Ctrl+L",   "action": "QuickLoad"  }
    ],
    "World": [
      { "combo": "Space",    "action": "NextTurn"   },
      { "combo": "Alt+U",    "action": "ToggleUI"   }
    ],
    "Unit": [
      { "combo": "F",        "action": "Fortify"    },
      { "combo": "G",        "action": "Upgrade"    }
    ]
  }
}
Versioned so you can migrate old layouts automatically (e.g., add new actions with sane defaults).
Store on desktop at ~/.unciv/keybindings.json; on Android in files/.

3  Editing UI – “keyboard‑centric” but mobile‑aware
Tree on the left: Context → Action list.

Search bar top: fuzzy‑match action names & tooltips.

Bind‑capture panel right: press keys, hold modifiers, or tap “Record gesture” on mobile to capture a two‑finger tap, etc.

Conflict banner: “Alt + U is already used in Global → Toggle‑UI. Override?”

Profiles dropdown: Default, WASD, One‑handed, Gamepad‑A/B/X/Y… plus Export/Import buttons (handy for Steam Deck users).

4  Context stack & priority rules
kotlin
Copy
object InputRouter {
    private val ctxStack = ArrayDeque<Ctx>()   // top = most specific
    fun push(ctx: Ctx) = ctxStack.addFirst(ctx)
    fun pop(ctx: Ctx)  = ctxStack.removeFirst(ctx)
    fun onKey(combo: KeyCombo) {
        for (ctx in ctxStack) {
            findBinding(ctx, combo)?.let { fire(it.action); return }
        }
    }
}
The World Screen pushes Ctx.World; opening a City Screen pushes Ctx.City on top; pop on close.

Modal pop‑ups (Confirm / Cancel) push Ctx.Dialog, which shadows everything else so <kbd>Enter</kbd> never accidentally triggers End Turn.

A debug overlay or cheat console can push Ctx.Console with absolute priority.

5  Suggested first‑wave bindings (cover 80 % of requests so far)
Context	Action	Default combo	Rationale
Global	Toggle UI	Alt + U	Frequent for screenshots
Global	Cycle Windowed / Fullscreen	F11	Matches common engines
World	Next Turn	Space / Enter	Civ V behaviour
World	Quicksave / Quickload	F12 / Shift + F12	Steam screenshots moved to F13 on some laptops
World	Empire Overview	F2	Aligns with Civ cousins
World	Toggle Yields	Y	mnemonic
Unit	Fortify / Heal	F / H	mnemonic
Unit	Delete Unit	Del + Confirm	Safety double‑press
City	Buy Tile	B	mnemonic
City	Switch city L/R	← / →	Already half‑wired
Picker	Confirm / Cancel	Enter / Esc	Universal
Dialog	Yes / No	Y / N	Faster than arrow‑Enter

Feel free to tweak defaults; the point is to seed the config for easier discovery. 
[github.com](https://github.com/yairm210/Unciv/issues/8862)

6  Power‑user extras (v2 roadmap)
Chorded binds: detect sequences like <kbd>G</kbd> then <kbd>P</kbd> (“Go to Plot”) within 750 ms.

Modifier holds: hold <kbd>Shift</kbd> while right‑clicking a hex = queue multiple moves.

Macro recording: record a combo‑sequence and save as “Build‑Road‑Fortify” macro; expose in bindings like any other action.

On‑screen hints: HUD overlay shows the currently shadowed bindings in grey; helps new players learn muscle‑memory.

Accessibility layer: allow binding long‑press or double‑tap on mobile for any Action (important for motor‑impaired users).

7  Testing & regression safety net
Layer	Test type	Toolkit
JSON schema	Unit	kotlinx.serialization validation + JsonSchemaTest
Resolver	Property‑based	kotest‑property – generate random stacks & combos, assert single dispatch
UI	Screenshot diff	gdx‑tests headless with Robolectric GL stub
End‑to‑end	Scripted	expect scripts: load map → simulate <kbd>Space</kbd> 10× → assert year advanced 10 turns

Add CI job that loads player‑exported binding profiles found in ./bindings/ and validates there are no duplicate combos within each context.

8  Migration & rollout
Phase 0 (now) – skeleton API behind “Debug → Keybinding beta” toggle.

Phase 1 – ship World‑screen actions + JSON import/export; ask community to contribute profiles for AZERTY, QWERTZ, Dvorak, Colemak.

Phase 2 – port City, Picker, Dialog contexts; deprecate hard‑coded isKeyJustPressed() calls.

Phase 3 – enable on Android with long‑press and double‑tap gesture recording; announce in Play Store changelog.

Mic-drop summary
Turn one of Unciv’s roughest UX edges into a moddable, profile‑driven input system that rivals AAA 4X titles—without breaking Android or mod compatibility.

Implementing Actions + Bindings + Resolver once lets every screen, unit, and future mod share the same rock‑solid foundation—and makes the community your QA force by letting them export, swap, and perfect their own control schemes.

**Comment by yairm210** [comment]
...is that AI generated? Because it has double texts, doesn't explain what problem it intends to solve, has broken markdown, and references a whole bunch of things that are just not in this project.
No AI slop, please, life is hard enough as it is

**Comment by SomeTroglodyte** [comment]
> is that AI generated

Looked more like 'Prepared in some external editor with a very different markdown support' - then pasted twice losing some glue markup - to me. I was waiting for OP to notice and reformat, then to answer 'we already got that, got that, ... yes but we're limited by upstream frameworks' to some recognizable points. Of course, recordable macros will stay out-of-scope here. But I expect once it's readable there might be a good idea or two in there.

**Comment by yairm210** [comment]
Closing this as it's been implemented
If there's other stuff we'll discuss it specifically
